### PR TITLE
Phase 6: Asymmetric Fixed Asinh Scales (Pos/Neg Pressure)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current asinh transform `asinh(p * 0.75)` compresses positive and negative pressure equally. But airfoil pressure is physically asymmetric:
- **Negative pressure** (suction side): large magnitude (-5 to -10 Cp), highly concentrated at leading edge — needs stronger compression
- **Positive pressure** (stagnation): moderate magnitude, broader distribution — needs lighter compression

The single scale s=0.75 must compromise between these two regimes. Using separate fixed scales for positive and negative pressure allows optimal compression of each.

**CRITICAL DISTINCTION FROM DEAD END #2096 (learnable scale):** PR #2096 tried `nn.Parameter(0.75)` and it collapsed to s→0.003 (trivial shortcut via joint optimization). This proposal uses **fixed, non-learnable** hyperparameter values — no gradient descent, no collapse risk. The failure mode of #2096 does not apply.

**Physical reasoning for initial values:**
- `s_neg = 1.5` — stronger compression for large negative suction peaks (where model underpredicts)
- `s_pos = 0.5` — lighter compression for moderate positive stagnation pressures

## Instructions

**Step 1: Add flags** to argparse:
```python
parser.add_argument('--asinh_asymmetric', action='store_true', default=False,
                    help='Use separate fixed asinh scales for pos/neg pressure')
parser.add_argument('--asinh_scale_neg', type=float, default=1.5,
                    help='Asinh scale for negative pressure (suction peaks)')
parser.add_argument('--asinh_scale_pos', type=float, default=0.5,
                    help='Asinh scale for positive pressure (stagnation)')
```
Add to Config dataclass:
```python
asinh_asymmetric: bool = False
asinh_scale_neg: float = 1.5
asinh_scale_pos: float = 0.5
```

**Step 2: Modify the asinh transform** (find where `asinh_pressure` is applied to both inputs and targets):

```python
if cfg.asinh_pressure:
    if cfg.asinh_asymmetric:
        p_channel = _yp[:, :, 2:3]  # pressure channel
        p_neg = torch.where(p_channel < 0,
                            torch.asinh(p_channel * cfg.asinh_scale_neg),
                            torch.zeros_like(p_channel))
        p_pos = torch.where(p_channel >= 0,
                            torch.asinh(p_channel * cfg.asinh_scale_pos),
                            torch.zeros_like(p_channel))
        _yp[:, :, 2:3] = p_neg + p_pos
    else:
        _yp[:, :, 2:3] = torch.asinh(_yp[:, :, 2:3] * cfg.asinh_scale)
```

**Apply the SAME transform to targets `_y` as well** — the transform must be symmetric between predictions and targets.

**Step 3: Update the inverse transform at evaluation time.** Find where `torch.sinh` is used to invert the pressure transform and apply the asymmetric inverse:
```python
if cfg.asinh_asymmetric:
    p_pred = pred[:, :, 2:3]
    p_neg = torch.where(p_pred < 0,
                        torch.sinh(p_pred) / cfg.asinh_scale_neg,
                        torch.zeros_like(p_pred))
    p_pos = torch.where(p_pred >= 0,
                        torch.sinh(p_pred) / cfg.asinh_scale_pos,
                        torch.zeros_like(p_pred))
    pred[:, :, 2:3] = p_neg + p_pos
```

**Experiment runs** (3 scale pairs, 2 seeds each):

```bash
# Trial 1: s_neg=1.5, s_pos=0.5, seed 42
python train.py --agent edward \
  --wandb_name "edward/asym-neg15-pos05-s42" \
  --wandb_group phase6/asymmetric-asinh \
  --asinh_pressure --asinh_asymmetric --asinh_scale_neg 1.5 --asinh_scale_pos 0.5 \
  --seed 42 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3

# Trial 1: seed 73
python train.py --agent edward \
  --wandb_name "edward/asym-neg15-pos05-s73" \
  --wandb_group phase6/asymmetric-asinh \
  --asinh_pressure --asinh_asymmetric --asinh_scale_neg 1.5 --asinh_scale_pos 0.5 \
  --seed 73 [same flags]

# Trial 2: moderate asymmetry, seed 42 + 73
python train.py --agent edward \
  --wandb_name "edward/asym-neg10-pos075-s42" \
  --wandb_group phase6/asymmetric-asinh \
  --asinh_pressure --asinh_asymmetric --asinh_scale_neg 1.0 --asinh_scale_pos 0.75 \
  --seed 42 [same flags]

python train.py --agent edward \
  --wandb_name "edward/asym-neg10-pos075-s73" \
  --wandb_group phase6/asymmetric-asinh \
  --asinh_pressure --asinh_asymmetric --asinh_scale_neg 1.0 --asinh_scale_pos 0.75 \
  --seed 73 [same flags]
```

W&B group: `phase6/asymmetric-asinh`

## Baseline

Current best (16-seed ensemble, PR #2093):

| Metric | Baseline |
|--------|----------|
| p_in   | **12.1** |
| p_oodc | **6.6**  |
| p_tan  | **29.1** |
| p_re   | **5.8**  |

Single-model references (with symmetric asinh s=0.75):
- seed=42: p_in≈12.71, p_oodc≈8.13
- seed=73: p_in≈12.2, p_oodc≈7.59

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent edward \
  --wandb_name "edward/baseline-s42" --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT edward:

## Results

### Implementation
Added `--asinh_asymmetric`, `--asinh_scale_neg`, `--asinh_scale_pos` flags. Refactored all 14 inline asinh transform sites into two helper functions (`_asinh_fwd`, `_asinh_inv`) for clean asymmetric support. The helpers dispatch to per-sign scaling when `asinh_asymmetric=True`, falling back to the original symmetric `asinh_scale` otherwise.

### Individual Run Results

| Config | Seed | p_in | p_oodc | p_tan | p_re | W&B Run |
|--------|------|------|--------|-------|------|---------|
| Strong (neg=1.5, pos=0.5) | 42 | 13.6 | 8.3 | 30.8 | 6.5 | 7uz2d0ol |
| Strong (neg=1.5, pos=0.5) | 73 | 13.2 | 8.2 | 30.6 | 6.7 | 1v86awoi |
| Moderate (neg=1.0, pos=0.75) | 42 | 13.4 | 7.9 | 30.6 | 6.5 | dj0elhj2 |
| Moderate (neg=1.0, pos=0.75) | 73 | 13.1 | 8.1 | 30.2 | 6.5 | rgs2r8jj |

### vs Baseline Single-Model Mean (p_in=13.03, p_oodc=7.83, p_tan=30.29, p_re=6.45)

| Config | p_in | p_oodc | p_tan | p_re |
|--------|------|--------|-------|------|
| Strong (1.5/0.5) mean | 13.40 (+2.8%) | 8.25 (+5.4%) | 30.70 (+1.4%) | 6.60 (+2.3%) |
| Moderate (1.0/0.75) mean | 13.25 (+1.7%) | 8.00 (+2.2%) | 30.40 (+0.4%) | 6.50 (+0.8%) |

### Command Used
```bash
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py --agent edward \
  --wandb_name "edward/asym-neg15-pos05-s42" \
  --wandb_group phase6/asymmetric-asinh \
  --asinh_pressure --asinh_asymmetric --asinh_scale_neg 1.5 --asinh_scale_pos 0.5 \
  --seed 42 [all baseline flags]
```

W&B group: `phase6/asymmetric-asinh`

### What happened

**Negative result.** Both asymmetric variants are worse than the symmetric baseline (s=0.75) on all metrics.

- **Strong asymmetry (neg=1.5, pos=0.5)**: Worst results, especially p_oodc (+5.4%). The aggressive negative compression (1.5) over-squashes suction peaks, making them harder for the model to distinguish. The light positive compression (0.5) under-compresses stagnation pressures.
- **Moderate asymmetry (neg=1.0, pos=0.75)**: Closer to baseline but still worse. The near-symmetric scales (1.0 vs 0.75) produce minimal differentiation while breaking the uniform compression that the model has been tuned for.
- **Key insight**: The symmetric `asinh(p * 0.75)` is already a good compromise. The model's capacity and training dynamics have adapted to this specific compression — breaking symmetry introduces a distributional mismatch at the zero-crossing boundary that hurts more than the per-regime optimization helps. The z-score normalization applied on top of asinh already handles scale differences between suction and stagnation regions.

### Code quality note
The refactoring from 14 inline `torch.asinh/sinh` calls to two helper functions (`_asinh_fwd`, `_asinh_inv`) is a net code quality improvement regardless of the experiment outcome — fewer duplication sites, easier to modify in future.

### Suggested follow-ups
- The symmetric scale s=0.75 appears to be a well-tuned sweet spot. Further asinh modifications are unlikely to yield gains.
- If pressure regime-specific handling is still desired, a learned per-region weighting in the loss function (not the target transform) might be a better approach — but previous learned-weight experiments have also failed (PR #2013, #2096).

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2108 because ADVISOR: Closing — **negative result confirmed.** Both asymmetric asinh configurations underperform the symmetric s=0.75 baseline on all four surface metrics:

| Config | p_in | p_oodc | p_tan | p_re |
|--------|------|--------|-------|------|
| Baseline (symmetric 0.75) | 13.03 | 7.83 | 30.29 | 6.45 |
| Strong (1.5/0.5) mean | 13.40 (+2.8%) | 8.25 (+5.4%) | 30.70 (+1.4%) | 6.60 (+2.3%) |
| Moderate (1.0/0.75) mean | 13.25 (+1.7%) | 8.00 (+2.2%) | 30.40 (+0.4%) | 6.50 (+0.8%) |

**Key insight:** The symmetric asinh(p×0.75) is already well-tuned. Breaking sign symmetry introduces distributional mismatch at the zero-crossing boundary that outweighs any per-regime benefit. The z-score normalization applied on top of asinh already handles scale differences between suction and stagnation regions.

**Adding to dead ends:** Asymmetric asinh scaling is a dead end. Combined with learnable scale (#2096) and asinh velocity (#2098), this closes the asinh transform direction entirely — the current s=0.75 is the right setting.

Good analysis on why the approach didn't work. The refactoring of 14 inline transform sites into helpers is a nice code quality improvement — if the helpers prove useful for future experiments, we'll incorporate them separately..
